### PR TITLE
[RFC][WIP] Full DTensor based FlexAttention

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.12.0) (oldest at bottom):
* __->__ #169393

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @Lucaskabela @wanchaol @tianyu-l @wz337 @XilunWu @d4l3k @pragupta @SherlockNoMad @ppwwyyxx

### CODE REVIEW

File: torch/distributed/tensor/_ops/_flex_attention.py [hunk 1]
```diff
@@ -0,0 +1,674 @@
...
+        assert grad_logsumexp.shape == logsumexp.shape
+        # TODO: we asume the grad_logsumexp are all zeros if it is not a DTensor,
+        # but I'm not sure if we can safely assume this.
+        for i, s in enumerate(logsumexp._local_tensor.shape):
+            grad_logsumexp = grad_logsumexp.narrow(i, 0, s)
+        grad_logsumexp = DTensor.from_local(
+            grad_logsumexp,
+            device_mesh=out.device_mesh,
+            placements=out.placements,
+        )
```
Comment (fegin) [inline torch/distributed/tensor/_ops/_flex_attention.py]: @drisspg

Comment (drisspg) [reply torch/distributed/tensor/_ops/_flex_attention.py]: Confirming; the behavior of autograd funcs for outputs that dont end up have a back propping grad is to allocate a zeros_like of the output, so should be good if it is not used in downstream ops

Comment (drisspg) [reply torch/distributed/tensor/_ops/_flex_attention.py]: https://docs.pytorch.org/docs/stable/generated/torch.autograd.function.FunctionCtx.set_materialize_grads.html

File: torch/distributed/tensor/_ops/_flex_attention.py [hunk 2]
```diff
@@ -0,0 +1,668 @@
...
+    """
+    Propagate sharding strategy for flex_attention and redistribute inputs.
+
+    This function performs custom sharding propagation for flex_attention
+    instead of using the standard propagator.propagate_op_sharding_non_cached()
+    because the standard path fails for the following reasons:
+
+    1. **Modified operator signature**: We modify the op signature to only
+       include the necessary mask tensors for sharding validation, since mask
+       tensors cannot be resharded. This simplification is needed because
```
Comment (tianyu-l) [inline torch/distributed/tensor/_ops/_flex_attention.py:299]: could you say more about "mask tensors cannot be resharded"?

I saw that `needs_redistribute` is still used normally to redistribute args (including mask tensors?)

File: torch/distributed/tensor/_ops/_flex_attention.py [hunk 3]
```diff
@@ -0,0 +1,668 @@
...
+    # An all-gather is needed to combine the Qs before the attention.
+    seq_dim_sharding: PlacementList = [
+        Shard(2),  # output,
+        Shard(2),  # logsumexp,
+        Shard(2),  # max_logits
+        Shard(2),  # q,
+        Replicate(),  # k,
+        Replicate(),  # v,
+        Shard(2),  # kv_num_blocks,
+        Shard(2),  # kv_indices,
```
Comment (tianyu-l) [inline torch/distributed/tensor/_ops/_flex_attention.py:113]: Is this intended? I forgot what they mean, but it sounds they should be `Replicate`.

File: torch/distributed/tensor/_api.py [hunk 4]
```diff
@@ -1374,3 +1381,255 @@ def zeros(  # type: ignore[no-untyped-def]
...
+                kv + kv_offset,
+            )
+
+        return offset_mask_mod
+
+    return create_mask_mod(*global_offset)
+
+
+def create_distributed_block_mask(
+    mask_mod: _mask_mod_signature,
```
Comment (tianyu-l) [inline torch/distributed/tensor/_api.py:1454]: I guess with CP (+ load balancing), users are still responsible for modifying the mask_mod to do (1) local to global q idx mapping and (2) position_ids / load balancer based idx permutation.

File: torch/distributed/tensor/_api.py [hunk 5]
```diff
@@ -1374,3 +1381,253 @@ def zeros(  # type: ignore[no-untyped-def]
         device_mesh=device_mesh,
         placements=placements,
     )
+
+
+def shard_mask_mod(
```
Comment (wanchaol) [inline torch/distributed/tensor/_api.py:1414]: I think these APIs should live in the flex attention module instead of the dtensor module, as it's very specific to flex attention

File: torch/_dynamo/variables/higher_order_ops.py [hunk 6]
```diff
@@ -3580,7 +3580,11 @@ def create_wrapped_node(
     ):
         from .._trace_wrapped_higher_order_op import TransformGetItemToIndex
 
-        def create_scalar():
+        def create_scalar(query):
```
Comment (drisspg) [inline torch/_dynamo/variables/higher_order_ops.py]: nit: for me -> we should be calling torch.empty here I remember this not working from ages ago and I dont know why lets re confirm.

File: torch/distributed/tensor/_api.py (hunk 5 above)
Comment (fegin) [reply torch/distributed/tensor/_api.py:1414]: Yes, we should probably move it out. But not flex attention but under distributed. Otherwise, there will be endless circular dependencies.

File: torch/distributed/tensor/_ops/_flex_attention.py [hunk 7]
```diff
@@ -0,0 +1,678 @@
...
+            "FlexAttention + DTensor doesn't support score_mod_other_buffers yet."
+        )
+
+    # Convert DTensor buffers in mask_mod_other_buffers to local tensors
+    mask_mod_other_buffers_local = pytree.tree_map(
+        lambda x: x._local_tensor if isinstance(x, DTensor) else x,
+        mask_mod_other_buffers,
+    )
+
+    if not isinstance(grad_logsumexp, DTensor):
```
Comment (fegin) [inline torch/distributed/tensor/_ops/_flex_attention.py:619]: @bdhirsh @drisspg

Comment (bdhirsh) [reply torch/distributed/tensor/_ops/_flex_attention.py:619]: @fegin I'm trying to run a repro to understand what's going on better. But running the tests in this PR errors for me with the below. Is there a repro somewhere I can go off of?
```
 python test/distributed/tensor/test_flex_attention.py -k test_flex_attention_replicate
 ...
   File "/data/users/hirsheybar/new2/pytorch/torch/_dynamo/variables/higher_order_ops.py", line 4145, in _call_function
    score_mod_node, score_mod_lifted_args = self.create_wrapped_node(
                                            ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/users/hirsheybar/new2/pytorch/torch/_dynamo/variables/higher_order_ops.py", line 4079, in create_wrapped_node
    bhmn = [create_scalar() for _ in range(4)]
            ^^^^^^^^^^^^^^^
  File "/data/users/hirsheybar/new2/pytorch/torch/_dynamo/variables/higher_order_ops.py", line 4065, in create_scalar
    if query.call_obj_hasattr(tx, "_local_tensor").value:
```

Comment (fegin) [reply torch/distributed/tensor/_ops/_flex_attention.py:619]: Sorry, forgot to upstream the latest version. Just updated the PR. You can reproduce the issue with 

`pytest  test/distributed/tensor/test_flex_attention.py -k test_flex_attention_replicate`
